Create observability.md (#808)
redis/docs: content/integrate/prometheus-with-redis-enterprise/observability.md

These operations are not necessarily problematic, but they can be if executed against data structures holding
a large number of elements (for example, a list with 1 million elements).

However, the [KEYS]({{< relref "/commands/keys" >}}) command should almost never be run against a
production system, since returning a list of all keys in a large Redis database can cause significant slowdowns
and block other operations. If you need to scan the keyspace, especially in a production cluster, always use the
[SCAN]({{< relref "/commands/scan" >}}) command instead.

### Troubleshooting

The best way to discover slow operations is to view the slow log.
The slow log is available in the Redis Enterprise and Redis Cloud consoles:
* [Redis Enterprise slow log docs](https://redis.io/docs/latest/operate/rs/clusters/logging/redis-slow-log/)
* [Redis Cloud slow log docs](https://redis.io/docs/latest/operate/rc/databases/view-edit-database/#other-actions-and-info)

Redis Cloud dashboard showing slow database operations
{{< image filename="/images/slow_log.png" alt="Redis Cloud dashboard showing slow database operations" >}}

| Issue | Remediation |
| ------ | :------ |

> |The KEYS command shows up in the slow log |Find the application that issues the KEYS command and replace it with a SCAN command. In an emergency situation, you can [alter the ACLs for the database user](https://redis.io/docs/latest/operate/rs/security/access-control/rbac/configure-acl/) so that Redis will reject the KEYS command altogether. |

Reviewer: Couldn't make a relref for this one because I couldn't find it. Don't know if it's moved or obsolete?

Author: good catch, fixed

|The slow log shows a significant number of slow, O(n) operations | If these operations are being issued against large data structures, then the application may need to be refactored to use more efficient Redis commands. |

> |The slow logs contains only O(1) commands, and these commands are taking several milliseconds or more to completen |This likely indicates that the database is underprovisioned. Consider increasing the number of shards and/or nodes. |

Reviewer: I think the word "completen" should exist, but AFAIK, it doesn't ;-) 
```suggestion
|The slow logs contains only O(1) commands, and these commands are taking several milliseconds or more to complete |This likely indicates that the database is underprovisioned. Consider increasing the number of shards and/or nodes. |
```

## Hot keys

A **hot key** is a key that is accessed extremely frequently (for example, thousands of times a second or more).

Each key in Redis belongs to one, and only one, shard.
For this reason, a hot key can cause high CPU utilization on that one shard,
which can increase latency for all other operations.

### Troubleshooting

You may suspect that you have a hot key if you see high CPU utilization on a single shard.
There are two main way to identify hot keys: using the Redis CLI and sampling the operations against Redis.

To use the Redis CLI to identify hot keys:

1. First confirm that you have enough available memory to enable an eviction policy.
2. Next, enable the LFU (least-frequently used) eviction policy on the database.
3. Finally, run `redis-cli --hotkeys`

You may also identify hot keys by sampling the operations against Redis.

> You can use do this by running the [MONITOR](https://redis.io/docs/latest/commands/monitor/) command

Reviewer: ```suggestion
You can do this by running the [MONITOR](https://redis.io/docs/latest/commands/monitor/) command
```

|The database in question will soon be unable to accept new data | round((bdb_used_memory/bdb_memory_limit) * 100) > 98 |
|The database in question will be unable to accept new data in two hours | round((bdb_used_memory/bdb_memory_limit) ** 100) < 98 and (predict_linear(bdb_used_memory[15m], 2 ** 3600) / bdb_memory_limit) > 0.3 and round(predict_linear(bdb_used_memory[15m], 2 * 3600)/bdb_memory_limit) > 0.98 |
|Database read operations are failing to find entries more than 50% of the time | (100 * bdb_read_hits)/(bdb_read_hits + bdb_read_misses) < 50 |
|In situations where TTL values are not set this indicates a problem | bdb_evicted_objects > 1 |
|Replication between nodes is not in a satisfactory state | bdb_replicaof_syncer_status > 0 |
|Record synchronization between nodes is not in a satisfactory state | bdb_crdt_syncer_status > 0 |
|The amount by which replication lags behind events is worrisome | bdb_replicaof_syncer_local_ingress_lag_time > 500 |
|The amount by which object replication lags behind events is worrisome | bdb_crdt_syncer_local_ingress_lag_time > 500 |

> |The expected number of active nodes is less than expected | count(node_up) != 3 |

Reviewer: ```suggestion
|The number of active nodes is less than expected | count(node_up) != 3 |
```

|Ephemeral storage will soon be exhausted | round((node_ephemeral_storage_free/node_ephemeral_storage_avail) * 100) <= 5 |
|The node in question is close to running out of memory | round((node_available_memory/node_free_memory) * 100) <= 15 |
|The node in question has exceeded expected levels of CPU usage | round((1 - node_cpu_idle) * 100) >= 80 |
|The shard in question is not reachable | redis_up == 0 |
|The master shard is not reachable | floor(redis_master_link_status{role="slave"}) < 1 |
|The shard in question has exceeded expected levels of CPU usage | redis_process_cpu_usage_percent >= 80 |
|The master shard has exceeded expected levels of CPU usage | redis_process_cpu_usage_percent{role="master"} > 0.75 and redis_process_cpu_usage_percent{role="master"} > on (bdb) group_left() (avg by (bdb)(redis_process_cpu_usage_percent{role="master"}) + on(bdb) 1.2 * stddev by (bdb) (redis_process_cpu_usage_percent{role="master"})) |
|The shard in question has an unhealthily high level of connections | redis_connected_clients > 500 |

## Appendix A: Grafana Dashboards

Grafana dashboards are available for Redis Enterprise Software and Redis Cloud deployments.

> These dashboards come in three styles, which may be used in concert with one another to provide
> a holistic picture of your deployment.

Reviewer: Bit flowery?
```suggestion
These dashboards come in three styles, which may be used together to provide
a full picture of your deployment.
```

1. Classic dashboards provide detailed information about the cluster, nodes, and individual databases.
2. Basic dashboards provide a high-level overviews of the various cluster components.

> 3. Extended dashboards which requires a third-party library to perform ReST calls.

Reviewer: ```suggestion
3. Extended dashboards. These require a third-party library to perform ReST calls.
```

> There are two additional sets of dashboards for Redis Enterprise software that provide drill-down functionality: the workflow dashboards.

Reviewer: ```suggestion
There are also two workflow dashboards for Redis Enterprise software that provide drill-down functionality.
```
